### PR TITLE
translation: EC legacy keys, handle OSSL_PKEY_PARAM_EC_PUB_X,Y requests

### PR DESCRIPTION
**Pull request by ldts**
Required by tpm2-tss to load legacy EC keys using the OpenSSL engine.

Fixes: https://github.com/tpm2-software/tpm2-tss/issues/2581

##### Checklist

- [ ] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by t8m** [comment]
It would be nice to have a test for this, would you be able to create one?

IMO it should be possible (in a test that can call internal functions) to downgrade a key to a legacy one and then retrieve the values.

File: crypto/evp/ctrl_params_translate.c [hunk 1]
```diff
@@ -1642,6 +1642,71 @@ static int get_payload_public_key(enum state state,
     return ret;
 }
 
+static int get_payload_public_key_ec(enum state state,
+                                     const struct translation_st *translation,
+                                     struct translation_ctx_st *ctx)
+{
+#ifndef OPENSSL_NO_EC
+    EVP_PKEY* pkey = ctx->p2;
```
[the remark this answers is not in the document]
Comment (ldts) [reply crypto/evp/ctrl_params_translate.c:1650]: um bloody editor, must have beautified it again on saving. ok sorry again about this, will fix.

yes will be adding a test on Monday

**Comment by ldts** [comment]
@t8m thanks. I had to refresh the PR as I noticed a small leak in the test code from some leftover code.

**Comment by t8m** [comment]
> @t8m thanks. I had to refresh the PR as I noticed a small leak in the test code from some leftover code.

No, the assign function makes the eckey be owned by the pkey so the fixup removing the NULL assignment was wrong.

**Comment by ldts** [comment]
Um, will test with the tip of the tree...on my development environment everything seemed fine:

```
jramirez@tpm-22:~/Src/import/openssl$ make TESTS="test_evp_extra" test

[...] 

    ok 30 - test_EC_priv_pub
    ok 31 - test_evp_get_ec_pub
    ok 32 - test_EC_priv_only_legacy
```

**Comment by ldts** [comment]
@t8m are all tests supposed to always pass?

File: crypto/evp/ctrl_params_translate.c [hunk 2]
```diff
@@ -1642,6 +1642,60 @@ static int get_payload_public_key(enum state state,
...
+    BIGNUM *x = NULL;
+    BIGNUM *y = NULL;
+    int ret = 0;
+
+    if (bnctx == NULL)
+        return 0;
+
+    ctx->p2 = NULL;
+
+    if (EVP_PKEY_get_base_id(pkey) != EVP_PKEY_EC) {
```
Comment (slontis) [inline crypto/evp/ctrl_params_translate.c:1664]: this check can be just be eckey == NULL

**Comment by t8m** [comment]
> @t8m are all tests supposed to always pass?

Yes, apart from intermittent failures due to network connectivity or other similar problems.

**Comment by t8m** [comment]
You have some memory leaks there.

File: test/evp_extra_test.c [hunk 3]
```diff
@@ -1031,6 +1031,64 @@ static int test_EC_priv_pub(void)
...
+    pad[0] = ec_pub[0];
+    BN_bn2bin(x, &pad[1]);
+    BN_bn2bin(y, &pad[33]);
+    if (!TEST_true(memcmp(ec_pub, pad, sizeof(ec_pub)) == 0))
+        goto err;
+
+    ret = 1;
+
+err:
+    OSSL_PARAM_BLD_free(bld);
```
Comment (t8m) [inline test/evp_extra_test.c:1083]: You're missing `OSSL_PARAM_free(params);` call here.

Comment (slontis) [reply test/evp_extra_test.c:1083]: Not sure what the buildtest one is..

Comment (slontis) [reply test/evp_extra_test.c:1083]: libctx not being passed somewhere

Comment (ldts) [reply test/evp_extra_test.c:1084]: @slontis what I just pushed fixes 

`./config --banner=Configured no-asm no-makedepend enable-buildtest-c++ enable-fips --strict-warnings -D_DEFAULT_SOURCE`

but is there a way for me to check all these configs locally without having to push to github to trigger?

**Comment by ldts** [comment]
@t8m @slontis dont know why that particular test fails but seems unrelated to this PR so will leave it at that

File: test/evp_extra_test.c (hunk 3 above)
Comment (slontis) [reply test/evp_extra_test.c:1084]: I normally just copy the config and build them separately if trying to track down a particular problem

**Comment by slontis** [comment]
You are correct, it is not related to your PR. Just some intermittent problem with windows buildbot.

File: test/evp_extra_test.c [hunk 4]
```diff
@@ -1094,6 +1153,80 @@ static int test_EC_priv_only_legacy(void)
...
+    EC_KEY *eckey = NULL;
+    BIGNUM *priv = NULL;
+    BIGNUM *x = NULL;
+    BIGNUM *y = NULL;
+    int ret = 0;
+
+    if (!TEST_ptr(libctx = OSSL_LIB_CTX_new()))
+        goto err;
+
+    if (!TEST_ptr(oldctx = OSSL_LIB_CTX_set0_default(libctx)))
```
Comment (slontis) [inline test/evp_extra_test.c:1172]: Can I ask why you need to make the call to set0_default() ?

Comment (slontis) [reply test/evp_extra_test.c:1172]: There is a EC_KEY_new_by_curve_name_ex()

Comment (slontis) [reply test/evp_extra_test.c:1172]: (Only set0_default() as a last resort... I see it as a hack)

Comment (ldts) [reply test/evp_extra_test.c:1172]: no reason really, will fix - I thought it was cleaner this way (the function you pointed too is not being used much) but sure, will fix

**Comment by ldts** [comment]
hi, anything else pending?

**Review by slontis — APPROVED** [review]
LGTM

**Review by t8m — APPROVED** [review]

**Comment by slontis** [comment]
@t8m can you reapprove..

**Comment by slontis** [comment]
windows buildbot failures are not related to this PR.

**Comment by ldts** [comment]
do I need to add your tags (acked, reviewed-by?) to the commit?

**Comment by ldts** [comment]
> 24 hours has passed since 'approval: done' was set, but as this PR has been updated in that time the label 'approval: ready to merge' is not being automatically set. Please review the updates and set the label manually.

from where I am sitting this is clear as mud :) but please ping me if I need to do something to get this merged

**Comment by t8m** [comment]
> @slontis @t8m do I need to add your tags (acked, reviewed-by?) to the commit?

Nope, we'll add them when merging.

**Pull request closed without merging**